<a href="https://colab.research.google.com/github/elfyld/Kod-devleri/blob/main/Pure_For_Multiple_Runs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from google.colab import drive
from os.path import join

drive.mount('./drive', force_remount=True)
path_prefix = './drive/My Drive'

from google.colab import drive
drive.mount('/content/drive')



Mounted at ./drive
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [94]:
protein_name="1zg4" #
fname='consurf.csv' #excelde dosyanızı kaydettiğiniz ad
df = pd.read_csv(join(path_prefix, fname))
df

POS SEQ    3LATOM  ...  CONFIDENCE INTERVAL COLORS  MSA DATA RESIDUE VARIETY
0      1   H   HIS26:A  ...                         8,6      4/69             N,H
1      2   P   PRO27:A  ...                         9,7     19/69               P
2      3   E   GLU28:A  ...                         6,2     19/69         D,E,Q,A
3      4   T   THR29:A  ...                         6,1     19/69         V,T,P,A
4      5   L   LEU30:A  ...                         8,5     20/69             V,L
..   ...  ..       ...  ...                         ...       ...             ...
258  259   L  LEU286:A  ...                         7,5     36/69           I,V,L
259  260   I  ILE287:A  ...                         5,1     33/69     T,V,I,F,M,A
260  261   K  LYS288:A  ...                         7,4     28/69           H,E,K
261  262   H  HIS289:A  ...                         7,2     18/69           E,H,V
262  263   W  TRP290:A  ...                         7,4     18/69           F,Y,W

[263 rows x 9 columns]

In [ ]:
df['COLOR']= df.COLOR.astype(int)


In [ ]:
Conserved_Amino_acids=pd.DataFrame(columns=['Amino Acids','Position','Color'])
Pos=[]
Seq=[]
Color=[]
copy=df.copy()
for i in df.index:
  if(7<=copy['COLOR'][i]<=9):
    Pos.append(copy['POS'][i])
    Seq.append(copy['SEQ'][i])
    Color.append(copy['COLOR'][i])

Conserved_Amino_acids['Amino Acids']=Seq
Conserved_Amino_acids['Position']=Pos
Conserved_Amino_acids['Color']=Color

Conserved_Amino_acids.sort_values(by='Position')

Amino Acids  Position  Color
0            H         1      7
1            P         2      8
2            E        12      9
3            A        17      8
4            R        18      7
..         ...       ...    ...
82           A       222      9
83           A       223      7
84           S       231      8
85           N       249      7
86           A       253      7

[87 rows x 3 columns]

In [ ]:
Conserved_Amino_acids['Color'].describe()

count    87.000000
mean      7.919540
std       0.852062
min       7.000000
25%       7.000000
50%       8.000000
75%       9.000000
max       9.000000
Name: Color, dtype: float64

In [ ]:
Position=[]
Sequence=[]
ColorNum=[]
for i in df.index:
  if(1<=df['COLOR'][i]<=6):
   Position.append(df['POS'][i])
   ColorNum.append(df['COLOR'][i])
   Sequence.append(df['SEQ'][i])

Unconserved_Amino_acids=pd.DataFrame(columns=['Amino Acids','Position','Color'])

Unconserved_Amino_acids['Amino Acids']= Sequence
Unconserved_Amino_acids['Position']=Position
Unconserved_Amino_acids['Color']=ColorNum
Unconserved_Amino_acids


Amino Acids  Position  Color
0             E         3      4
1             T         4      3
2             L         5      6
3             V         6      6
4             K         7      3
..          ...       ...    ...
171           L       259      5
172           I       260      2
173           K       261      5
174           H       262      4
175           W       263      5

[176 rows x 3 columns]

In [ ]:
aa_names = {'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',
         'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', 
         'GLY': 'G', 'HIS': 'H', 'HSD': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', 
         'ALA': 'A', 'VAL':'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'}
mutation_list = ["ALA","ARG","ASN","ASP","CYS","GLU","GLN","GLY","HIS","ILE",
                 "LEU","LYS","MET","PHE","PRO","SER","THR","TRP","TYR","VAL"]


**Function for mutating conserved amino acids**

In [ ]:
# mutation on %5 of conserved amino acids
import random as rnd
def Conserved_Mut(mutP,ProteinL):
  check=1
  M_Conserved=Conserved_Amino_acids.copy()
  row,col=Conserved_Amino_acids.shape
  while(check<263*ProteinL/100):  #263 ü kendi proteninizin uzunluğu ile değiştirirn
    ranom_index=rnd.randint(0,row-1)
    random_mut=aa_names[mutation_list[rnd.randint(0,len(mutation_list)-1)]]

    if(random_mut!=M_Conserved.at[ranom_index,'Amino Acids']):
      M_Conserved.at[ranom_index,'Amino Acids']=random_mut
    
      check+=1

  return  M_Conserved 

**Function for mutating unconserved amino acids**

In [ ]:
# mutation on %5 of unconserved amino acids
def Unconserved_mut(mutP,ProteinL):
  check=1
  M_Unconserved=Unconserved_Amino_acids.copy()
  row,col=Unconserved_Amino_acids.shape
  while(check<ProteinL*mutP/100):#263 ü kendi proteninizin uzunluğu ile değiştirirn
    ranom_index=rnd.randint(0,row-1)
    random_mut=aa_names[mutation_list[rnd.randint(0,len(mutation_list)-1)]]

    if(random_mut!=M_Unconserved.at[ranom_index,'Amino Acids']):
      M_Unconserved.at[ranom_index,'Amino Acids']=random_mut
    
      check+=1
  return 



**Mutations on both conserved and uncorved areas of the protein**





In [101]:
# mutation on %5 of conserved and unconserved amino acids
mutated_Seq=""
f= open("Mixed mutation sequences.txt","w")
e = 1
for i in range(10):
  frmes= [Conserved_Mut(5,263),Unconserved_mut(5,263)]
  x=pd.concat(frmes,ignore_index=True)
  x.sort_values(by='Position')
  for i in x['Amino Acids'].values:
      mutated_Seq+=i
  
  f.write(">" + str(e)+"_"+protein_name+"_%5 mixed"+"\n"+ mutated_Seq+"\n")
  e+=1
  mutated_Seq=""
  
f.close() 

**%10 Mutation on only conserved areas**

In [102]:
mutated_Seq=""
f= open("% 10 conserved area muatations.txt","w")
e = 1
for i in range(10):
  frmes= [Conserved_Mut(10,263),Unconserved_Amino_acids.copy()]
  x=pd.concat(frmes,ignore_index=True)
  x.sort_values(by='Position')
  for i in x['Amino Acids'].values:
      mutated_Seq+=i
  
  f.write(">" + str(e)+"_"+protein_name+"_%10 conserved"+"\n"+ mutated_Seq+"\n")#burayı değiştirip fasta formatı yapabilir misin? Buralarda sekansları bi txt file a yazıyor
  e+=1
  mutated_Seq=""
  
f.close() 

**%10 Mutation on only unconserved areas**

In [103]:
mutated_Seq=""
f= open("%10 unconserved area muatations.txt","w")
c=1
for i in range(10):
  frmes= [Conserved_Amino_acids.copy(),Unconserved_mut(10,263)]
  x=pd.concat(frmes,ignore_index=True)
  x.sort_values(by='Position')
  for i in x['Amino Acids'].values:
      mutated_Seq+=i
  
  f.write(">" + str(c)+"_"+protein_name+"_%10 unconserved"  +"\n"+ mutated_Seq+"\n") #burayı değiştirip fasta formatı yapabilir misin? Buralarda sekansları bi txt file a yazıyor
  c+=1
  mutated_Seq=""
  
f.close() 

*%10 muatation on whole protein**

In [104]:
seq=df.copy()
row,col=Unconserved_Amino_acids.shape
f= open("%10 whole protein.txt","w")
k = 1
for i in range(10):
  check=1
  mut=""
  while(check<263*10/100):
    ranom_index=rnd.randint(0,row-1)
    random_mut=aa_names[mutation_list[rnd.randint(0,len(mutation_list)-1)]]

    if(random_mut!=seq.at[ranom_index,'SEQ']):
      seq.at[ranom_index,'SEQ']=random_mut
    
      check+=1
  for i in seq['SEQ'].values:
      mut+=i
  f.write(">"+ str(k)+"_"+protein_name+"_%10" + "\n" + mut +"\n") #burayı değiştirip fasta formatı yapabilir misin? Buralarda sekansları bi txt file a yazıyor
  k+=1
f.close()

From 263 colums of amino acids 176 of it unconserved and 87 conserved

Which means %66.92015209
 of the protein is unconserved
            %33.07984791 of the protein is conserved